ocaml/tests/test_repository_helpers.ml

```diff
@@ -0,0 +1,1218 @@
+    `QuickAndAutoDocumented
+      [
+        ( ( ("1.2.3", "3.el7"), ("gt", ("1.2.3", "4.el7")) ), Ok false )
+      ; ( ( ("1.2.3", "3.el7"), ("gt", ("1.2.4", "3.el7")) ), Ok false )
+      ; ( ( ("1.2.3", "3.el7"), ("gt", ("1.2.3", "2.el7")) ), Ok true )
+      ; ( ( ("1.2.3", "3.el7"), ("gt", ("1.2.2", "3.el7")) ), Ok true )
+
+      ; ( ( ("1.2.3", "3.el7"), ("lt", ("1.2.3", "2.el7")) ), Ok false )
+      ; ( ( ("1.2.3", "3.el7"), ("lt", ("1.2.2", "3.el7")) ), Ok false )
+      ; ( ( ("1.2.3", "3.el7"), ("lt", ("1.2.3", "4.el7")) ), Ok true )
+      ; ( ( ("1.2.3", "3.el7"), ("lt", ("1.2.4", "3.el7")) ), Ok true )
+
+      ; ( ( ("1.2.3", "3.el7"), ("eq", ("1.2.3", "3.el7")) ), Ok true )
+      ; ( ( ("1.2.3", "3.el7"), ("eq", ("1.2.4", "3.el7")) ), Ok false)
+
+      ; ( ( ("1.2.3", "3.el7"), ("gte", ("1.2.3", "4.el7")) ), Ok false )
+      ; ( ( ("1.2.3", "3.el7"), ("gte", ("1.2.4", "3.el7")) ), Ok false )
+      ; ( ( ("1.2.3", "3.el7"), ("gte", ("1.2.3", "3.el7")) ), Ok true )
+      ; ( ( ("1.2.3", "3.el7"), ("gte", ("1.2.3", "2.el7")) ), Ok true )
+      ; ( ( ("1.2.3", "3.el7"), ("gte", ("1.2.2", "3.el7")) ), Ok true )
+
+      ; ( ( ("1.2.3", "3.el7"), ("lte", ("1.2.3", "2.el7")) ), Ok false )
+      ; ( ( ("1.2.3", "3.el7"), ("lte", ("1.2.2", "3.el7")) ), Ok false )
+      ; ( ( ("1.2.3", "3.el7"), ("lte", ("1.2.3", "3.el7")) ), Ok true )
+      ; ( ( ("1.2.3", "3.el7"), ("lte", ("1.2.3", "4.el7")) ), Ok true )
+      ; ( ( ("1.2.3", "3.el7"), ("lte", ("1.2.4", "3.el7")) ), Ok true )
+
+      ; ( ( ("1.2.3", "3.el7"), ("let", ("1.2.3", "3.el7")) ),
+          Error Applicability.Invalid_inequality)
+      ]
+end)
+
+module UpdateInfoMetaDataOfXml = Generic.MakeStateless (struct
+  module Io = struct
+    type input_t = string
+
+    type output_t = (UpdateInfoMetaData.t, exn) result
+
+    let string_of_input_t x = x
+
+    let fields = Fmt.Dump.([
+        field "checksum" (fun (r:UpdateInfoMetaData.t) -> r.checksum) string
+      ; field "location" (fun (r:UpdateInfoMetaData.t) -> r.location) string
+      ])
+
+    let string_of_output_t =
+      Fmt.(str "%a" Dump.(result ~ok:(record @@ fields) ~error:exn))
+  end
+
+  let checksum = "123abc"
+
+  let location = "repodata/123abc.xml.gz"
+
+  let normal =
+      {|<checksum type="sha256">|} ^ checksum ^ {|</checksum>|}
+    ^ {|<location href="|} ^ location ^ {|"/>|}
+
+  let missing_checksum = {|<location href="|} ^ location ^ {|"/>|}
+
+  let missing_href =
+      {|<checksum type="sha256">|} ^ checksum ^ {|</checksum>|}
+    ^ {|<location hef="|} ^ location ^ {|"/>"|}
+
+  let with_repomd s = {|<?xml version="1.0" encoding="UTF-8"?>|} ^ "<repomd>" ^ s ^ "</repomd>"
+
+  let with_updateinfo s = {|<data type="updateinfo">|} ^ s ^ {|</data>|}
+
+  let updateinfo = with_updateinfo normal
+
+  let transform input =
+    try Ok (UpdateInfoMetaData.of_xml (Xml.parse_string input))
+    with e -> Error e
+
+  let tests =
+    `QuickAndAutoDocumented
+      [ (with_repomd "",
```

Comment:
Similarly to the json tests above, I think it'd be easier to read this test if we were testing a single xml string for each test case, e.g.

```
let bad_test_cases = [
  ({|<?xml version="1.0" encoding="UTF-8"?><repomd></repomd>|}, invalid_repomd_xml ;
 ...
]
```

Reply:
Thanks. Just want to make my understanding clear: do you mean following example would be better?

```
[ ({|<?xml version="1.0" encoding="UTF-8"?>
     <repomd>
     </repomd>
    |},
   Error Api_errors.(Server_error (invalid_repomd_xml, [])))
; ({|<?xml version="1.0" encoding="UTF-8"?>
     <repomd>
       <data type="primary"></data>
     </repomd>
    |},
   Error Api_errors.(Server_error (invalid_repomd_xml, [])))`
]
```
And the point is: it would be easier to read the test cases with complete XML/JSON string embedded in each case. And this overwhelms benefit of reducing the duplication of those XML/JSON strings.

Reply:
Yes, that's what I meant! I also mentioned this alternative:

> If there a lots of xml test string, you could put the strings in their own files

By that, I meant have a folder for each type of xml you want to test, e.g.

```
├── centaurus
│   ├── invalid_repomd_xml
│   │   ├── 1.xml
│   │   ├── 2.xml
│   │   └── 3.xml
│   ├── valid_repomd_xml
│   │   ├── 1.xml
│   │   ├── 2.xml
```

and then your ocaml tests become something like this

```
let test () =
  Sys.readdir "ocaml/tests/centaurus/invalid_repomd_xml" |> Array.map (fun fname -> Unixext.string_of_file fname |> try_parse |> assert_error) ;
  Sys.readdir "ocaml/tests/centaurus/valid_repomd_xml" |> Array.map (fun fname -> Unixext.string_of_file fname |> try_parse |> assert_ok)
```

I think I have a preference for this latter way, but it's up to you